File: app/controllers/StockSentiment.scala
package controllers

import javax.inject._

import play.api.{Configuration, Logger}
import play.api.libs.json.{JsObject, JsString, JsValue, Json}
import play.api.libs.ws._
import play.api.mvc._

import scala.concurrent.{ExecutionContext, Future}

@Singleton
class StockSentiment @Inject()(ws: WSClient,
                               configuration: Configuration,
                               cc: ControllerComponents)
                              (implicit ec: ExecutionContext)
  extends AbstractController(cc) {

  private val logger = Logger(this.getClass)

  private val sentimentUrl = configuration.get[String]("sentiment.url")

  private val tweetUrl = configuration.get[String]("tweet.url")

  case class Tweet(text: String)
  
  private implicit val tweetReads = Json.reads[Tweet]

  def get(symbol: String): Action[AnyContent] = Action.async {
    logger.info(s"getting stock sentiment for $symbol")

    val futureStockSentiments: Future[Result] = for {
      tweets <- getTweets(symbol) // get tweets that contain the stock symbol
      futureSentiments = loadSentimentFromTweets(tweets.json) // queue web requests each tweets' sentiments
      sentiments <- Future.sequence(futureSentiments) // when the sentiment responses arrive, set them
    } yield Ok(sentimentJson(sentiments))

    futureStockSentiments.recover {
      case nsee: NoSuchElementException =>
        InternalServerError(Json.obj("error" -> JsString("Could not fetch the tweets")))
    }
  }

  private def getTextSentiment(text: String): Future[WSResponse] = {
    logger.info(s"getTextSentiment: text = $text")

    ws.url(sentimentUrl).post(Map("text" -> Seq(text)))
  }

  private def getAverageSentiment(responses: Seq[WSResponse], label: String): Double = {
    responses.map { response =>
      (response.json \\ label).head.as[Double]
    }.sum / responses.length.max(1)
  } // avoid division by zero

  private def loadSentimentFromTweets(json: JsValue): Seq[Future[WSResponse]] = {
    (json \ "statuses").as[Seq[Tweet]] map (tweet => getTextSentiment(tweet.text))
  }

  private def getTweets(symbol:String): Future[WSResponse] = {
    logger.info(s"getTweets: symbol = $symbol")

    ws.url(tweetUrl.format(symbol)).get.withFilter { response =>
      response.status == OK
    }
  }

  private def sentimentJson(sentiments: Seq[WSResponse]): JsObject = {
    logger.info(s"sentimentJson: sentiments = $sentiments")

    val neg = getAverageSentiment(sentiments, "neg")
    val neutral = getAverageSentiment(sentiments, "neutral")
    val pos = getAverageSentiment(sentiments, "pos")

    val response = Json.obj(
      "probability" -> Json.obj(
        "neg" -> neg,
        "neutral" -> neutral,
        "pos" -> pos
      )
    )

    val classification =
      if (neutral > 0.5)
        "neutral"
      else if (neg > pos)
        "neg"
      else
        "pos"

    val r = response + ("label" -> JsString(classification))
    logger.info(s"response = $r")

    r
  }

}

File: app/controllers/StockSentiment.java
package controllers;

import com.fasterxml.jackson.databind.JsonNode;
import play.Configuration;
import play.libs.Json;
import play.libs.concurrent.Futures;
import play.libs.concurrent.HttpExecutionContext;
import play.libs.ws.WSClient;
import play.libs.ws.WSResponse;
import play.mvc.Controller;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.stream.Collectors.averagingDouble;
import static java.util.stream.Collectors.toList;
import static utils.Streams.stream;

@Singleton
public class StockSentiment extends Controller {

    private WSClient wsClient;
    private Configuration configuration;
    private HttpExecutionContext ec;

    @Inject
    public StockSentiment(WSClient wsClient, Configuration configuration, HttpExecutionContext ec) {
        this.wsClient = wsClient;
        this.configuration = configuration;
        this.ec = ec;
    }

    public CompletionStage<Result> get(String symbol) {
        return fetchTweets(symbol)
                .thenComposeAsync(this::fetchSentiments)
                .thenApplyAsync(this::averageSentiment)
                .<Result>thenApplyAsync(Results::ok)
                .exceptionally(this::errorResponse);
    }

    private CompletionStage<List<String>> fetchTweets(String symbol) {

        final String tweetUrl = configuration.getString("tweet.url");
        final CompletionStage<WSResponse> futureResponse = wsClient.url(tweetUrl)
                .setQueryParameter("q", "$" + symbol)
                .get();

        final CompletionStage<WSResponse> filter = futureResponse.thenApplyAsync(response -> {
            if (response.getStatus() == Http.Status.OK) {
                return response;
            } else {
                return null;
            }
        }, ec.current());

        return filter.thenApplyAsync(response -> {
            final List<String> statuses = stream(response.asJson().findPath("statuses"))
                    .map(s -> s.findValue("text").asText())
                    .collect(Collectors.toList());
            return statuses;
        });
    }

    private CompletionStage<List<JsonNode>> fetchSentiments(List<String> tweets) {
        String url = configuration.getString("sentiment.url");
        Stream<CompletionStage<WSResponse>> sentiments = tweets.stream().map(text -> {
            return wsClient.url(url).post("text=" + text);
        });
        return Futures.sequence(sentiments::iterator).thenApplyAsync(this::responsesAsJson);
    }

    private List<JsonNode> responsesAsJson(List<WSResponse> responses) {
        return responses.stream().map(WSResponse::asJson).collect(toList());
    }

    private JsonNode averageSentiment(List<JsonNode> sentiments) {
        double neg = collectAverage(sentiments, "neg");
        double neutral = collectAverage(sentiments, "neutral");
        double pos = collectAverage(sentiments, "pos");

        String label = (neutral > 0.5) ? "neutral" : (neg > pos) ? "neg" : "pos";

        return Json.newObject()
                .put("label", label)
                .set("probability", Json.newObject()
                        .put("neg", neg)
                        .put("neutral", neutral)
                        .put("pos", pos));
    }

    private double collectAverage(List<JsonNode> jsons, String label) {
        return jsons.stream().collect(averagingDouble(json -> json.findValue(label).asDouble()));
    }

    private Result errorResponse(Throwable ignored) {
        return internalServerError(Json.newObject().put("error", "Could not fetch the tweets"));
    }
}
